src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +216,249 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
+                String fieldName = "m" + count++;
+                var mt = methodType(m.getReturnType(), JLRA.getExecutableSharedParameterTypes(m));
+                var thrown = JLRA.getExecutableSharedExceptionTypes(m);
+                var exceptionTypeDescs =
+                     thrown.length == 0 ? DEFAULT_RETHROWS
+                                        : Stream.concat(DEFAULT_RETHROWS.stream(),
+                                                        Arrays.stream(thrown).map(MethodHandleProxies::desc))
+                                                .distinct().toList();
+                methods.add(new MethodInfo(desc(mt), exceptionTypeDescs, fieldName));
+
+                // find the types referenced by this method
+                addElementType(referencedTypes, m.getReturnType());
+                addElementTypes(referencedTypes, JLRA.getExecutableSharedParameterTypes(m));
+                addElementTypes(referencedTypes, JLRA.getExecutableSharedExceptionTypes(m));
+            }
+
+            if (uniqueName == null)
+                throw newIllegalArgumentException("no method in ", intfc.getName());
+
+            // create a dynamic module for each proxy class, which needs access
+            // to the types referenced by the members of the interface including
+            // the parameter types, return type and exception types
+            var loader = intfc.getClassLoader();
+            Module targetModule = newDynamicModule(loader, referencedTypes);
+
+            // generate a class file in the package of the dynamic module
+            String pn = targetModule.getName();
+            String n = intfc.getName();
+            int i = n.lastIndexOf('.');
+            String cn = i > 0 ? pn + "." + n.substring(i + 1) : pn + "." + n;
+            byte[] template = createTemplate(loader, ClassDesc.of(cn), desc(intfc), uniqueName, methods);
+            // define the dynamic module to the class loader of the interface
+            var definer = new Lookup(intfc).makeHiddenClassDefiner(cn, template, Set.of(), DUMPER);
+
+            @SuppressWarnings("removal")
+            var sm = System.getSecurityManager();
+            Lookup lookup;
+            if (sm != null) {
+                PrivilegedAction<Lookup> pa =  () -> definer.defineClassAsLookup(true);
+                @SuppressWarnings("removal")
+                var l = AccessController.doPrivileged(pa);
+                lookup = l;
+            } else {
+                lookup = definer.defineClassAsLookup(true);
+            }
+            WRAPPER_TYPES.add(lookup.lookupClass());
+            return new WeakReference<>(lookup);
+        }
+    };
+
+    private static Lookup getProxyClassLookup(Class<?> intfc) {
+        WeakReference<Lookup> r = PROXY_LOOKUPS.get(intfc);
+        if (r.refersTo(null)) {
+            PROXY_LOOKUPS.remove(intfc);
+            r = PROXY_LOOKUPS.get(intfc);
+        }
+        return r.get();
```

Comment:
Sorry I re-read Chen's question and missing the last half of the sentence.   My reply was about:  If the referent is not GC'ed, `r.get()` returns non-null which will be returned to the caller.  If the referent becomes weakly reachable, `r.get()` will return null.

OK.  In any case, I was thinking to use a local variable to keep the lookup object strongly reachable.   I adjusted the patch to keep a class value of a holder of `WeakReference` ([ab3d8b2](https://github.com/mlchung/jdk/commit/ab3d8b2f086dfac4f8fda1ddce415270c0c53d2b)):

```
    private static Lookup getProxyClassLookup(Class<?> intfc) {
        WeakReferenceHolder<Lookup> r = PROXY_LOOKUPS.get(intfc);
        Lookup lookup = r.get();
        if (lookup == null) {
             // If the referent is cleared, create a new value and update cached weak reference.
            lookup = newProxyLookup(intfc);
            r.set(lookup);
        }
        return lookup;
    }
```

Reply:
Ok, thanks. This is similar to the solution I would have suggested.

The only thing I'm unsure about is races. ClassValue allows multiple threads to race to compute the final value, but in the end only 1 is chosen that is returned to all threads.

If multiple threads race to create a new proxy lookup here, they will all get a distinct lookup returned to them. I think this is okay? Though, it would mean we have less sharing of proxy classes.

Maybe this code should synchronize on `r` to make sure only 1 lookup is created per interface. (Letting threads race to compute the value is also a possibility, but it also leads some threads to do work that is then discarded).

Reply:
> If multiple threads race to create a new proxy lookup here, they will all get a distinct lookup returned to them. I think this is okay? Though, it would mean we have less sharing of proxy classes.

yes, that's okay.  We can also do what you suggest to synchronize on `r`.

Reply:
I don't have a strong opinion, but I'm in favor of synchronizing (in a double-checked locking style), because the amount of work that would be discarded in the case of a race would be non-trivial.